machine_learning_hep/processerdhadrons_mult.py

```diff
@@ -105,14 +109,41 @@ def gethistonormforselevt_mult(self, df_evt, dfevtevtsel, label, var, weightfunc
         df_no_vtx = df_to_keep[~tag_vtx.values]
         # events with reco zvtx > 10 cm after previous selection
         df_bit_zvtx_gr10 = filter_bit_df(df_to_keep, 'is_ev_rej', [[3], [1, 2, 7, 12]])
-        if weightfunc is not None:
-            weightssel = evaluate(weightfunc, dfevtevtsel[var])
+        if useweightfromfunc is not None:
+            weightssel = None
+            weightsnovtx = None
+            weightsgr10 = None
+            if useweightfromfunc is True:
+                weightssel = evaluate(self.weightfunc, dfevtevtsel[var])
+                weightsnovtx = evaluate(self.weightfunc, df_no_vtx[var])
+                weightsgr10 = evaluate(self.weightfunc, df_bit_zvtx_gr10[var])
+            else:
+                weightssel = []
+                weightsnovtx = []
+                weightsgr10 = []
+                for iw in dfevtevtsel[var]:
```

Comment:
This looks good, but may I suggest that it would be worth it to have a function that takes care of the business of weight estimation, something like:
```
def make_weights(df):
            weight_list = []
            if useweightfromfunc is True:
                weight_list = evaluate(self.weightfunc, df)
            else:
                 for iw in df: 
                    value = self.weighthist.GetBinContent(self.weighthist.FindBin(iw))
                    # warning, the histogram has empty bins at high mult.
                    # (>125 ntrkl) so a check is needed to avoid a 1/0 division
                    # when computing the inverse of the weight
                    if value == 0:
                        value = 1.
                    weight_list.append(value)
```

Reply:
I guess that makes sense. @arthur-gal just let me know, I can do it quickly

Reply:
@benedikt-voelkel  @njacazio, thanks Nicolo, yes the code is a bit redundant, using a function is clearly a best option. I let you handle it Benedikt, thanks